core/src/main/scala/kafka/log/remote/RemoteLogManager.scala

```diff
@@ -237,25 +238,39 @@ class RemoteLogManager(fetchLog: TopicPartition => Option[Log],
     // Compact topics and internal topics are filtered here as they are not supported with tiered storage.
     def nonSupported(partition: Partition): Boolean = {
       Topic.isInternal(partition.topic) ||
-        partition.log.exists(log => log.config.compact || !log.config.remoteStorageEnable) ||
+        // fetchLog should reflect up to date configuration
+        fetchLog(partition.topicPartition).exists(log => log.config.compact || !log.config.remoteStorageEnable) ||
         partition.topicPartition.topic().equals(TopicBasedRemoteLogMetadataManagerConfig.REMOTE_LOG_METADATA_TOPIC_NAME)
     }
 
-    val leaderPartitions = partitionsBecomeLeader.filterNot(nonSupported)
-      .map(p => new TopicIdPartition(topicIds.get(p.topic), p.topicPartition) -> p).toMap
-    val followerPartitions = partitionsBecomeFollower.filterNot(nonSupported)
-      .map(p => new TopicIdPartition(topicIds.get(p.topic), p.topicPartition))
-    if (leaderPartitions.nonEmpty || followerPartitions.nonEmpty) {
-      debug(s"Effective topic partitions after filtering compact and internal topics, " +
-        s"leaders: ${leaderPartitions.keySet} and followers: $followerPartitions")
-      topicIds.forEach((topic, uuid) => this.topicIds.put(topic, uuid))
-      remoteLogMetadataManager.onPartitionLeadershipChanges(leaderPartitions.keySet.asJava, followerPartitions.asJava)
-      followerPartitions.foreach {
-        topicIdPartition => doHandleLeaderOrFollowerPartitions(topicIdPartition, _.convertToFollower())
-      }
-      leaderPartitions.foreach {
-        case (topicIdPartition, partition) =>
-          doHandleLeaderOrFollowerPartitions(topicIdPartition, _.convertToLeader(partition.getLeaderEpoch))
+    /**
+     * The following scenarios are possible when the onLeadershipChange method is invoked:
+     *
+     * 1. LeaderAndIsrRequest arrives before ConfigHandler config change for remoteStorageEnable: In this case, there will be a separate call for config change.
+     * 2. LeaderAndIsrRequest arrives after ConfigHandler config for remoteStorageEnable: In this case, the config changes will not be lost as there is a fetchLog call now to fetch the log updates.
+     * 3. LeaderAndIsrRequest arrives at same time as ConfigHandler config for remoteStorageEnable: Falls in to 1 or 2 above depending on sequencing
+     * 
+     * This method is safe for all the three scenarios and will be able to pick up the updated configuration.
```

Comment:
This is my understanding of the issue.

Either LeaderAndIsrRequest comes in first, or the Config change comes in first. If LeaderAndIsrRequest comes in first, we are happy because the first call will add the topicId to the map, and then RLM will be definitely initialized on the second call.

If ConfigHandler gets called first before LeaderAndIsr is received, and the broker doesn't yet have a topicId, then we can simply ignore the call in `maybeBootstrapRemoteLogComponents`. We will just update the log config to say RLM is enabled. Now, when the LeaderAndIsr request comes in, the config changes will not be lost and we will initialize RLM for the topic.

I have added this little check in our LinkedIn branch for this corner case.
https://github.com/linkedin/kafka/blob/3.0-li-rc1/core/src/main/scala/kafka/server/ConfigHandler.scala#L83-L87

Can you please explain if that check won't suffice for some corner case?

Reply:
Hi Tirtha
I hope your question was answered by Kamal's comment below? The summary is that we won't get a LISR request in cases when TS is enabled dynamically and hence can't rely on them to update the topicID, hence, we changed the dependency to zookeeper directly.